File: DFO.m
clear
% INITIALISE PARAMETERS (N: swarm size, D: dimensionality)
N=100; D = 30; delta = 0.001; maxIter=1000;
% LOWER AND UPPER BOUNDS
lowerB(1,1:D)=-5.12; upperB(1,1:D)=5.12;
s = 0;			% INITIAL INDEX OF BEST FLY
X = zeros(N,D);	% FLIES MATRIX
fitness(1,1:N) = realmax; % FITNESS VALUES 

for i=1:N
	for d=1:D
		X(i,d) = lowerB(d)+rand()*(upperB(d)-lowerB(d));
	end
end

for itr=1:maxIter
	% EVALUATE FITNESS OF THE FLIES AND FIND THE BEST
	fitness = f(X);
	[sFitness, s] = min(fitness);
	disp(['Iteration: ' , num2str(itr),'    Best fly index: ', num2str(s), '    Fitness value: ', num2str(sFitness), ] ) 

	% UPDATE EACH FLY INDIVIDUALLY
	for i=1:N
		% ELITIST STRATEGY (DON'T UPDATE BEST FLY)
		if i == s continue; end

		% FIND BEST NEIGHBOUR FOR EACH FLY
		left=mod(i-2,N)+1; right=mod(i,N)+1; % INDICES: LEFT & RIGHT FLIES
		if fitness(right)<fitness(left) bNeighbour = right; 
		else bNeighbour = left; end	

		for d=1:D
			% DISTURBANCE MECHANISM
			if rand()<delta X(i,d) = lowerB(d)+rand()*(upperB(d)-lowerB(d)); continue; end

			% UPDATE EQUATION
			X(i,d) = X(bNeighbour,d) + rand()*( X(s,d)- X(i,d) );

			% OUT OF BOUND CONTROL
			if or( X(i,d) < lowerB(d), X(i,d) > upperB(d) )
				X(i,d) = lowerB(d)+rand()*(upperB(d)-lowerB(d));
			end
		end
	end
end

% EVALUATE FITNESS OF THE FLIES AND FIND THE BEST
fitness = f(X);
[sFitness, s] = min(fitness);

disp( ['Final best fitness=', num2str(sFitness)] ) 

File: DFO.py
import numpy as np

# FITNESS FUNCTION (SPHERE FUNCTION)
def f(x): # x IS A VECTOR REPRESENTING ONE FLY
	sum = 0.0
	for i in range(len(x)):
		sum = sum + np.power(x[i],2)
	return sum

N = 100					# POPULATION SIZE
D = 30					# DIMENSIONALITY 
delta = 0.001			# DISTURBANCE THRESHOLD 
maxIterations = 1000	# ITERATIONS ALLOWED
lowerB = [-5.12]*D		# LOWER BOUND (IN ALL DIMENSIONS)
upperB = [ 5.12]*D		# UPPER BOUND (IN ALL DIMENSIONS)

# INITIALISATION PHASE
X = np.empty([N,D]) # EMPTY FLIES ARRAY OF SIZE: (N,D)
fitness = [None]*N  # EMPTY FITNESS ARRAY OF SIZE N

# INITIALISE FLIES WITHIN BOUNDS
for i in range(N):
	for d in range(D):
		X[i,d] = np.random.uniform(lowerB[d], upperB[d])
		
# MAIN DFO LOOP
for itr in range (maxIterations):
	for i in range(N): # EVALUATION
		fitness[i] = f(X[i,])
	s = np.argmin(fitness) # FIND BEST FLY

	if (itr%100 == 0): # PRINT BEST FLY EVERY 100 ITERATIONS
		print ("Iteration:", itr, "\tBest fly index:", s, 
			   "\tFitness value:", fitness[s])

	# TAKE EACH FLY INDIVIDUALLY 
	for i in range(N): 
		if i == s: continue # ELITIST STRATEGY

		# FIND BEST NEIGHBOUR
		left = (i-1)%N
		right = (i+1)%N
		bNeighbour = right if fitness[right]<fitness[left] else left

		for d in range(D): # UPDATE EACH DIMENSION SEPARATELY 
			if (np.random.rand() < delta):
				X[i,d] = np.random.uniform(lowerB[d], upperB[d])
				continue;

			u = np.random.rand()
			X[i,d] = X[bNeighbour,d] + u*(X[s,d] - X[i,d])

			# OUT OF BOUND CONTROL
			if X[i,d] < lowerB[d] or X[i,d] > upperB[d]:
				X[i,d] = np.random.uniform(lowerB[d], upperB[d])

for i in range(N): fitness[i] = f(X[i,]) # EVALUATION
s = np.argmin(fitness) # FIND BEST FLY

print("\nFinal best fitness:\t", fitness[s])
print("\nBest fly position:\n",  X[s,])	
